src/XSAlgo/XSAlgo_ShapeProcessor.cxx

```diff
@@ -0,0 +1,548 @@
+  Handle(ShapeFix_Edge) anEdgeFixer = new ShapeFix_Edge;
+  anEdgeFixer->FixSameParameter(aTmpEdge);
+  Standard_Real aTolerance = BRep_Tool::Tolerance(aTmpEdge);		       
+  Standard_Boolean aSameRangeFlag = BRep_Tool::SameRange(aTmpEdge);
+  Standard_Boolean aSameParameterFlag = BRep_Tool::SameParameter(aTmpEdge);
+
+  // if result is not nice, try to call projection and take the best
+  if (aTolerance > Min(1., 2. * thePrecision) || !aSameRangeFlag)
+  {
+    //pdn trying to recompute pcurve 
+    TopoDS_Edge anEdgePr = MakeEdgeOnCurve(theEdge);
+    anEdgeFixer->FixAddPCurve(anEdgePr, theFace, theIsSeam, thePrecision);
+    anEdgeFixer->FixSameParameter(anEdgePr);
+    const Standard_Real aTolerancePr = BRep_Tool::Tolerance(anEdgePr);
+    //pdn choose the best pcurve
+    if (aTolerancePr < aTolerance || !aSameRangeFlag)
+    {
+      aSameRangeFlag = BRep_Tool::SameRange(anEdgePr);
+      aSameParameterFlag = BRep_Tool::SameParameter(anEdgePr);
+      aTolerance = aTolerancePr;
+      aTmpEdge = anEdgePr;
+    }
+  }
+
+  // get corrected pcurve from the temporary edge, and put to original
+  anEdgeAnalyzer.PCurve(aTmpEdge, theFace, aCurve2D, aCurve2DParam1, aCurve2DParam2, Standard_False);
+  if (theIsSeam)
+  {
+    Standard_Real aReversedTmpEdgeParam1;
+    Standard_Real aReversedTmpEdgeParam2;
+    TopoDS_Edge aReversedTmpEdge = TopoDS::Edge(aTmpEdge.Reversed());
+    anEdgeAnalyzer.PCurve(aReversedTmpEdge,
+                          theFace,
+                          aSeamPCurve,
+                          aReversedTmpEdgeParam1,
+                          aReversedTmpEdgeParam2,
+                          Standard_False);
+    if (theEdge.Orientation() == TopAbs_REVERSED) //:abv 14.11.01: coneEl.sat loop
+    {
+      aBuilder.UpdateEdge(theEdge, aSeamPCurve, aCurve2D, theFace, aTolerance);
+    }
+    else
+    {
+      aBuilder.UpdateEdge(theEdge, aCurve2D, aSeamPCurve, theFace, aTolerance);
+    }
+  }
+  else
+  {
+    aBuilder.UpdateEdge(theEdge, aCurve2D, theFace, aTolerance);
+  }
+
+  aBuilder.UpdateVertex(aVertex1, aTolerance);
+  aBuilder.UpdateVertex(aVertex2, aTolerance);
+  aBuilder.Range(theEdge, theFace, aCurve2DParam1, aCurve2DParam2);
+  if (BRep_Tool::SameRange(theEdge))
+  {
+    aBuilder.SameRange(theEdge, aSameRangeFlag);
+  }
+  if (BRep_Tool::SameParameter(theEdge))
+  {
+    aBuilder.SameParameter(theEdge, aSameParameterFlag);
+  }
+
+  return Standard_True;
+}
+
+//=======================================================================
+//function : ToParameterMap
+//purpose  : 
+//=======================================================================
+XSAlgo_ShapeProcessor::ParameterMapPtr XSAlgo_ShapeProcessor::ToParameterMap(const DE_ShapeFixParameters& theParameters)
+{
+  // Helper lambda to convert enum to string.
+  auto makeString = [](const DE_ShapeFixParameters::FixMode theMode)
+    {
+      return std::to_string(static_cast<std::underlying_type<DE_ShapeFixParameters::FixMode>::type>(theMode));
```

Comment:
IS this magic c++11 support? Or we need to use macros?

Reply:
I think, everything used in this line of code is C++11 compatible. More specifically:
1. `std::underlying_type<T>` provides a member typedef type that names the underlying type of T. Note the I use here specifically `std::underlying_type<T>::type` ad not `std::underlying_type_t<T>`. The former was introduced in C++11, while latter was introduced in C++14.
2. `std::to_string()` - All overloads of this function were introduced in in C++11.